### PR TITLE
Add http 1.0 test

### PR DESCRIPTION
**Pull request by howardjohn**
This adds testing for https://github.com/istio/istio/pull/11511. The test is simple, and just sends a lot of http 1.0 requests to fortio. The client does not use fortio due to https://github.com/fortio/fortio/issues/324. If/when that is figured out we can switch curl out, but for now it does the job.

### CODE REVIEW

**Comment by ldemailly** [comment]
you shouldn't use a different client for http1.0 (see https://github.com/fortio/fortio/issues/324#issuecomment-462700414 ) - let me know if you need anything else

**Comment by ldemailly** [comment]
in other words, if you are trying to test for the most basic http 1.0 client; you should indeed use `fortio load/curl -http1.0` which will not have a `Host:` header like those clients

I suppose an http1.0 with host header test could be done too but that's basically http1.1 with `-halfclose`

**Comment by howardjohn** [comment]
> in other words, if you are trying to test for the most basic http 1.0 clients; you should indeed use `fortio load/curl -http1.0` which will not have a `Host:` header like those clients and like the standard allows/defines.
> 
> I suppose an http1.0 with host header test could be done too but that's basically http1.1 with `-halfclose`

Envoy [doesn't support http1.0 without a host header](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/protocol.proto#core-http1protocoloptions) so thats why we need it. 

I am running into some strange behavior with fortio though
* `time fortio curl -http1.0  -H "Host: fortio-server" fortio-server:8080/echo` = 1s
* `time fortio curl fortio-server:8080/echo` = 0.1s
* `time curl --http1.0  -v -H "Host: fortio-server" fortio-server:8080/echo` = 0.1s

For some reason fortio http1.0 requests are taking a full second. This only happens when running in kubernetes/istio - locally it works fast as expected, so I am not sure where the problem lies. But, because the requests are slow, we can't get the QPS desired (https://github.com/fortio/fortio/issues/327).

**Comment by ldemailly** [comment]
you just need to set `default_host_for_http_10` if I read the link above correctly? 

about your other problem, I’m (wild) guessing envoy or something is waiting for some timeout to process the request. try using -loglevel debug for fortio and for all the proxies involved and check where is the 1s spent ?
also try swapping client/server one at a time

**Comment by howardjohn** [comment]
```
23:07:37 D http_client.go:671> Read ok 143 total 143 so far (-0 headers = 143 data) HTTP/1.0 200 OK\r\ncontent-length: 0\r\ndate: Fri, 15 Feb 2019 23:07:36 GMT\r\nx-envoy-upstream-service-time: 1\r\nserver: envoy\r\nconnection: close\r\n\r\n
TIME SPENT HERE
23:07:38 D http_client.go:815> Closed ok &{{0xc00008e280}} from 10.35.247.122:8080 after reading 143 bytes
```

So it looks like we get a response from envoy and then spend a lot of time - not sure what exactly the time is spent doing.

**Comment by ldemailly** [comment]
interesting and if you run tcpdump or check envoy side, did envoy already close the connection right away, or did it wait 1s to do so ? I am guessing it didn't close yet

In http 1.0 mode, fortio, like simple clients, doesn't read the content-length header, it waits for EOF as per the standard - so I'm guessing that is what is going on and for some reason envoy doesn't immediately close that socket -- cc @PiotrSikora is that how envoy behaves in 1.0 mode?

**Comment by costinm** [comment]
I think "Connection: Close" header is needed in some implementations. Don't remember - but it's important to check what clients are doing ( and maybe find which are still doing 1.0 - php, python ?)

**Comment by ldemailly** [comment]
I do remember as I wrote several clients and servers back in 1.0 time. No headers are required from the client beside body length for a post, and a 1.0 server will close the socket immediately unless the non standard keep-alive is requested explicitly.

**Comment by costinm** [comment]
Of course, the shell fix can be done in separate PR.

File: perf/istio/http10/values.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+fortioImage: fortio/fortio:latest
```
Comment (costinm) [inline perf/istio/http10/values.yaml]: FYI: helm has a nice way to specify defaults ( | .... ).

**Comment by howardjohn** [comment]
> interesting and if you run tcpdump or check envoy side, did envoy already close the connection right away, or did it wait 1s to do so ? I am guessing it didn't close yet
> 
> In http 1.0 mode, fortio, like simple clients, doesn't read the content-length header, it waits for EOF as per the standard - so I'm guessing that is what is going on and for some reason envoy doesn't immediately close that socket -- cc @PiotrSikora is that how envoy behaves in 1.0 mode?

@ldemailly I ran a tcpdump on both sides, everything completes immediately. 

From server:
```
22:01:02.911602 IP 10.32.36.30.51332 > fortio-server-59685477bf-n726q.8080: Flags [P.], seq 2263:3017, ack 391, win 247, options [nop,nop,TS val 3993569575 ecr 4271217512], length 754: HTTP: GET / HTTP/1.1
22:01:02.912935 IP fortio-server-59685477bf-n726q.8080 > 10.32.36.30.51332: Flags [P.], seq 391:521, ack 3017, win 268, options [nop,nop,TS val 4271223310 ecr 3993569575], length 130: HTTP: HTTP/1.1 200 OK
22:01:02.913857 IP 10.32.36.30.51332 > fortio-server-59685477bf-n726q.8080: Flags [.], ack 521, win 256, options [nop,nop,TS val 3993569578 ecr 4271223310], length 0
```

From client:
```
22:01:02.910803 IP shell-5c6bbcc4c6-m9vzt.51332 > 10.32.72.92.8080: Flags [P.], seq 2263:3017, ack 391, win 247, options [nop,nop,TS val 3993569575 ecr 4271217512], length 754: HTTP: GET / HTTP/1.1
22:01:02.913661 IP 10.32.72.92.8080 > shell-5c6bbcc4c6-m9vzt.51332: Flags [P.], seq 391:521, ack 3017, win 268, options [nop,nop,TS val 4271223310 ecr 3993569575], length 130: HTTP: HTTP/1.1 200 OK
22:01:02.913685 IP shell-5c6bbcc4c6-m9vzt.51332 > 10.32.72.92.8080: Flags [.], ack 521, win 256, options [nop,nop,TS val 3993569578 ecr 4271223310], length 0
```

Request completes at `22:01:03.915875926` (.9s after all tcp action).

**Comment by ldemailly** [comment]
interesting (and unexpected) - what's the easiest way to reproduce this ? (ideally with just some static envoy config and no istio nor k8s but a k8s+istio simple setup is ok too)

**Comment by howardjohn** [comment]
There is probably a simpler way, but the way I did it: Run Istio and two fortio pods with annotation `HTTP10=1` (just like in this PR). Then just `fortio curl -http1.0 -H "Host: fortio" fortio:8080`.

**Comment by ldemailly** [comment]
afaik your tcpdump output is missing the FIN (or RST) btw... can you make sure you have everything ?

there should be some [F] or [R]

**Comment by howardjohn** [comment]
I am not very familiar with TCP/tcpdump, but I am pretty certain it was just those 3 lines each - I tried multiple times. I should have done curl and compared the dumps, I'll do that tomorrow maybe.

**Comment by ldemailly** [comment]
this is what a complete tcpdump would look like:
```
# tcpdump -i lo 
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on lo, link-type EN10MB (Ethernet), capture size 262144 bytes
17:07:29.351404 IP6 localhost.41672 > localhost.http-alt: Flags [S], seq 2750948181, win 43690, options [mss 65476,sackOK,TS val 20737 ecr 0,nop,wscale 7], length 0
17:07:29.351432 IP6 localhost.http-alt > localhost.41672: Flags [S.], seq 4129849749, ack 2750948182, win 43690, options [mss 65476,sackOK,TS val 20737 ecr 20737,nop,wscale 7], length 0
17:07:29.351453 IP6 localhost.41672 > localhost.http-alt: Flags [.], ack 1, win 342, options [nop,nop,TS val 20737 ecr 20737], length 0
17:07:29.352457 IP6 localhost.41672 > localhost.http-alt: Flags [P.], seq 1:60, ack 1, win 342, options [nop,nop,TS val 20738 ecr 20737], length 59: HTTP: GET / HTTP/1.0
17:07:29.352493 IP6 localhost.http-alt > localhost.41672: Flags [.], ack 60, win 342, options [nop,nop,TS val 20738 ecr 20738], length 0
17:07:29.352715 IP6 localhost.http-alt > localhost.41672: Flags [P.], seq 1:76, ack 60, win 342, options [nop,nop,TS val 20738 ecr 20738], length 75: HTTP: HTTP/1.0 200 OK
17:07:29.352764 IP6 localhost.http-alt > localhost.41672: Flags [F.], seq 76, ack 60, win 342, options [nop,nop,TS val 20738 ecr 20738], length 0
17:07:29.353175 IP6 localhost.41672 > localhost.http-alt: Flags [.], ack 76, win 342, options [nop,nop,TS val 20738 ecr 20738], length 0
17:07:29.353291 IP6 localhost.41672 > localhost.http-alt: Flags [F.], seq 60, ack 77, win 342, options [nop,nop,TS val 20738 ecr 20738], length 0
17:07:29.353307 IP6 localhost.http-alt > localhost.41672: Flags [.], ack 61, win 342, options [nop,nop,TS val 20738 ecr 20738], length 0
```

**Comment by ldemailly** [comment]
also your dump isn't of fortio -http1.0 (it says GET ... HTTP/**1.1**) unlike mine which has `GET / HTTP/1.0`

**Comment by howardjohn** [comment]
I figured out exactly what causes it.

Switching the Service from
```
  ports:
  - name: http
    port: 8080
```
to
```
  ports:
  - port: 8080
```

And it behaves as expected. I haven't yet looked into why though

**Comment by ldemailly** [comment]
that’s because without http envoy becomes a tcp proxy. can you get the actual tcpdump or get someone to confirm envoy’s behavior of not closing the socket immediately ?

**Comment by howardjohn** [comment]
Without `name: http` and using http1.0
```
19:17:27.844620 IP 10.32.2.82.35770 > fortio-server-59685477bf-zmlch.8080: Flags [S], seq 3457343744, win 28400, options [mss 1420,sackOK,TS val 503485082 ecr 0,nop,wscale 7], length 0
19:17:27.844659 IP fortio-server-59685477bf-zmlch.8080 > 10.32.2.82.35770: Flags [S.], seq 930713960, ack 3457343745, win 28160, options [mss 1420,sackOK,TS val 3813366528 ecr 503485082,nop,wscale 7], length 0
19:17:27.844806 IP 10.32.2.82.35770 > fortio-server-59685477bf-zmlch.8080: Flags [.], ack 1, win 222, options [nop,nop,TS val 503485082 ecr 3813366528], length 0
19:17:27.844921 IP 10.32.2.82.35770 > fortio-server-59685477bf-zmlch.8080: Flags [P.], seq 1:60, ack 1, win 222, options [nop,nop,TS val 503485083 ecr 3813366528], length 59: HTTP: GET / HTTP/1.0
19:17:27.844931 IP fortio-server-59685477bf-zmlch.8080 > 10.32.2.82.35770: Flags [.], ack 60, win 220, options [nop,nop,TS val 3813366528 ecr 503485083], length 0
19:17:27.845417 IP fortio-server-59685477bf-zmlch.8080 > 10.32.2.82.35770: Flags [P.], seq 1:76, ack 60, win 220, options [nop,nop,TS val 3813366528 ecr 503485083], length 75: HTTP: HTTP/1.0 200 OK
19:17:27.845449 IP fortio-server-59685477bf-zmlch.8080 > 10.32.2.82.35770: Flags [F.], seq 76, ack 60, win 220, options [nop,nop,TS val 3813366528 ecr 503485083], length 0
19:17:27.845604 IP 10.32.2.82.35770 > fortio-server-59685477bf-zmlch.8080: Flags [.], ack 76, win 222, options [nop,nop,TS val 503485083 ecr 3813366528], length 0
19:17:27.846185 IP 10.32.2.82.35770 > fortio-server-59685477bf-zmlch.8080: Flags [F.], seq 60, ack 77, win 222, options [nop,nop,TS val 503485084 ecr 3813366528], length 0
19:17:27.846203 IP fortio-server-59685477bf-zmlch.8080 > 10.32.2.82.35770: Flags [.], ack 61, win 220, options [nop,nop,TS val 3813366529 ecr 503485084], length 0
```

With `name: http` and using http1.0 - note it shows 1.1 here, but I passed `-http1.0` flag
```
19:21:47.843733 IP 10.32.2.82.36456 > fortio-server-59685477bf-zmlch.8080: Flags [P.], seq 2267:3021, ack 391, win 247, options [nop,nop,TS val 503745060 ecr 3813623705], length 754: HTTP: GET / HTTP/1.1
19:21:47.844717 IP fortio-server-59685477bf-zmlch.8080 > 10.32.2.82.36456: Flags [P.], seq 391:521, ack 3021, win 268, options [nop,nop,TS val 3813626507 ecr 503745060], length 130: HTTP: HTTP/1.1 200 OK
19:21:47.844902 IP 10.32.2.82.36456 > fortio-server-59685477bf-zmlch.8080: Flags [.], ack 521, win 256, options [nop,nop,TS val 503745061 ecr 3813626507], length 0
```

With `name: http` and not using http 1.0
```
19:20:14.344820 IP 10.32.2.82.36010 > fortio-server-59685477bf-zmlch.8080: Flags [P.], seq 755:1514, ack 131, win 231, options [nop,nop,TS val 503651568 ecr 3813448721], length 759: HTTP: GET / HTTP/1.1
19:20:14.345959 IP fortio-server-59685477bf-zmlch.8080 > 10.32.2.82.36010: Flags [P.], seq 131:261, ack 1514, win 244, options [nop,nop,TS val 3813533015 ecr 503651568], length 130: HTTP: HTTP/1.1 200 OK
19:20:14.346676 IP 10.32.2.82.36010 > fortio-server-59685477bf-zmlch.8080: Flags [.], ack 261, win 239, options [nop,nop,TS val 503651570 ecr 3813533015], length 0
```

**Comment by ldemailly** [comment]
this demonstrates you are looking at envoy-envoy traffic or envoy-destination and not client to envoy (which would have 1.0 and socket closing)

**Comment by howardjohn** [comment]
Ok so before I was looking at envoy-envoy, and it finishes right away. Now I have one of client to envoy

Fortio, showing the 1s delay
```
19:33:27.944004 IP 10.35.248.249.8080 > 10.32.2.82.59364: Flags [S.], seq 3258149487, ack 3450358404, win 43690, options [mss 65495,sackOK,TS val 3753766754 ecr 356620891,nop,wscale 7], length 0
19:33:27.944481 IP 10.35.248.249.8080 > 10.32.2.82.59364: Flags [.], ack 81, win 342, options [nop,nop,TS val 3753766754 ecr 356620891], length 0
19:33:27.948279 IP 10.35.248.249.8080 > 10.32.2.82.59364: Flags [P.], seq 1:144, ack 81, win 342, options [nop,nop,TS val 3753766758 ecr 356620891], length 143: HTTP: HTTP/1.0 200 OK
19:33:28.947563 IP 10.35.248.249.8080 > 10.32.2.82.59364: Flags [F.], seq 144, ack 81, win 342, options [nop,nop,TS val 3753767757 ecr 356620895], length 0
19:33:28.947736 IP 10.35.248.249.8080 > 10.32.2.82.59364: Flags [.], ack 82, win 342, options [nop,nop,TS val 3753767757 ecr 356621894], length 0
```

Curl, no delay
```
19:34:16.947898 IP 10.35.248.249.8080 > 10.32.2.82.59508: Flags [S.], seq 4213325499, ack 3301568370, win 43690, options [mss 65495,sackOK,TS val 3753815754 ecr 356669891,nop,wscale 7], length 0
19:34:16.948016 IP 10.35.248.249.8080 > 10.32.2.82.59508: Flags [.], ack 83, win 342, options [nop,nop,TS val 3753815754 ecr 356669891], length 0
19:34:16.952632 IP 10.35.248.249.8080 > 10.32.2.82.59508: Flags [P.], seq 1:144, ack 83, win 342, options [nop,nop,TS val 3753815758 ecr 356669891], length 143: HTTP: HTTP/1.0 200 OK
19:34:16.952865 IP 10.35.248.249.8080 > 10.32.2.82.59508: Flags [F.], seq 144, ack 84, win 342, options [nop,nop,TS val 3753815759 ecr 356669896], length 0
```

So the delay seems to be between envoy and the client. 

Thanks for all the help

**Comment by ldemailly** [comment]
yes so my diagnostic/guess earlier is correct, envoy when acting as an http 1.0 proxy holds off from closing the socket for 1 second, thus causing the delay.
(it doesn't affect curl because curl parses the content-length which is not required for 1.0 and closes the connection client side when it sees it has all the data, but again that's not how a normal/old style 1.0 client would behave, as fortio demonstrates)

**Comment by ldemailly** [comment]
so I strongly recommend you do use fortio -http1.0 for this test if you want it to most reflective of a likely client (not counting the point of fortio was to exercise everything istio may need without needing additional software)

**Comment by howardjohn** [comment]
From Envoy:
> See "delayed_close_timeout" https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto

**Pull request merged**